File: src/main/GO/leetcode/lists/topinterview/LeetCode341.go
package topinterview

type NestedIterator struct {
	vals []int
}

type NestedInteger struct {
}

func (this NestedInteger) IsInteger() bool {
	return false
}
func (this NestedInteger) GetInteger() int {
	return -1
}
func (this NestedInteger) GetList() []*NestedInteger {
	return nil
}

func ConstructorNestedInteger(nestedList []*NestedInteger) *NestedIterator {
	var vals []int
	var dfs func([]*NestedInteger)
	dfs = func(nestedList []*NestedInteger) {
		for _, nest := range nestedList {
			if nest.IsInteger() {
				vals = append(vals, nest.GetInteger())
			} else {
				dfs(nest.GetList())
			}
		}
	}
	dfs(nestedList)
	return &NestedIterator{vals}
}

func (this *NestedIterator) Next() int {
	val := this.vals[0]
	this.vals = this.vals[1:]
	return val
}

func (this *NestedIterator) HasNext() bool {
	return len(this.vals) > 0
}

File: src/main/java/leetcode/lists/topinterview/LeetCode341.java
package leetcode.lists.topinterview;

import java.util.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/10/7 00:13
 */
public class LeetCode341 {

    public class NestedIterator implements Iterator<Integer> {
        private final List<Integer> vals;
        private final Iterator<Integer> cur;

        public NestedIterator(List<NestedInteger> nestedList) {
            vals = new ArrayList<>();
            dfs(nestedList);
            cur = vals.iterator();
        }

        @Override
        public Integer next() {
            return cur.next();
        }

        @Override
        public boolean hasNext() {
            return cur.hasNext();
        }

        private void dfs(List<NestedInteger> nestedList) {
            for (NestedInteger nest : nestedList) {
                if (nest.isInteger()) {
                    vals.add(nest.getInteger());
                } else {
                    dfs(nest.getList());
                }
            }
        }
    }

    interface NestedInteger {
        boolean isInteger();

        Integer getInteger();

        List<NestedInteger> getList();
    }
}
